src/main/service/listManipulation/OpaPublicListService.ts

```diff
@@ -0,0 +1,114 @@
+                                const caseDetails = this.buildCaseDetails(hearing.case);
+                                rows.push({
+                                    ...caseDetails,
+                                    ...defendant,
+                                });
+                            });
+                        });
+                    });
+                });
+            });
+        });
+        return rows;
+    }
+
+    private processPartyRoles(hearing): any {
+        const defendants = [];
+        hearing.party?.forEach(party => {
+            if (party.partyRole === 'DEFENDANT') {
+                const defendant = this.buildDefendantDetails(party);
+                if (defendant) {
+                    defendants.push(defendant);
+                }
+            }
+        });
+        const prosecutor = opaPressListService.processProsecutor(hearing);
+        const allDefendants = [];
+        defendants.forEach(defendant => {
+            allDefendants.push({ ...defendant, prosecutor });
+        });
+        return allDefendants;
+    }
+
+    private buildDefendantDetails(party): any {
+        if (party.individualDetails) {
+            const individual = party.individualDetails;
+            return {
+                name: this.buildIndividualName(individual),
+                offence: this.buildOffences(individual),
+            };
+        } else if (party.organisationDetails) {
+            const organisation = party.organisationDetails;
+            return {
+                name: organisation.organisationName,
+                offence: this.buildOffences(organisation),
+            };
+        }
+        return null;
+    }
+
+    private buildIndividualName(individual) {
+        const firstName = ListParseHelperService.writeStringIfValid(individual?.individualFirstName);
+        const middleNames = ListParseHelperService.writeStringIfValid(individual?.individualMiddleName);
+        const surname = ListParseHelperService.writeStringIfValid(individual?.individualSurname);
+        const forenames = [firstName, middleNames].filter(n => n.length > 0).join(' ');
+        return [surname, forenames]
+            .filter(n => n.length > 0)
+            .join(', ')
+            .toString();
+    }
+
+    private buildOffences(defendant) {
+        const offences = [];
+        defendant.offence?.forEach(offence => {
+            offences.push(this.buildOffenceDetails(offence));
+        });
+        return offences;
+    }
+
+    private buildOffenceDetails(offence) {
+        return {
+            offenceTitle: ListParseHelperService.writeStringIfValid(offence.offenceTitle),
+            offenceSection: ListParseHelperService.writeStringIfValid(offence.offenceSection),
+            offenceReportingRestriction: offence.reportingRestrictionDetail?.filter(n => n.length > 0).join(', '),
+        };
+    }
+
+    private buildCaseDetails(hearingCase) {
```

Comment:
The logic here does not look correct. If there are multiple cases within the same hearing, currently you are looping through the case array and replacing the previous values with the last entry. You end up with just the record from the last case